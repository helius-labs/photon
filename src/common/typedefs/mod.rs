pub mod hash;
pub mod serializable_pubkey;
