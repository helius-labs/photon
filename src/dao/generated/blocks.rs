//! `SeaORM` Entity. Generated by sea-orm-codegen 0.10.6

use sea_orm::entity::prelude::*;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq)]
#[sea_orm(table_name = "blocks")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub slot: i64,
    pub parent_slot: i64,
    pub parent_blockhash: Vec<u8>,
    pub blockhash: Vec<u8>,
    pub block_height: i64,
    pub block_time: i64,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(has_many = "super::transactions::Entity")]
    Transactions,
}

impl Related<super::transactions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Transactions.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}
