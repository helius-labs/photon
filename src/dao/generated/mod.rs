//! `SeaORM` Entity. Generated by sea-orm-codegen 0.10.6

pub mod prelude;

pub mod account_transactions;
pub mod accounts;
pub mod blocks;
pub mod owner_balances;
pub mod state_trees;
pub mod token_accounts;
pub mod token_owner_balances;
pub mod transactions;
