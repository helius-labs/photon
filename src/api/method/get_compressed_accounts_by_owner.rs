use crate::dao::generated::accounts;
use sea_orm::{ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter, QueryOrder, QuerySelect};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use super::{
    super::error::PhotonApiError,
    utils::{Context, Limit, PAGE_LIMIT},
};
use crate::common::typedefs::{hash::Hash, serializable_pubkey::SerializablePubkey};

use super::utils::{parse_account_model, Account};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct Options {
    pub cursor: Option<Hash>,
    pub limit: Option<Limit>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetCompressedAccountsByOwnerRequest {
    pub owner: SerializablePubkey,
    pub cursor: Option<Hash>,
    pub limit: Option<Limit>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct PaginatedAccountList {
    pub items: Vec<Account>,
    pub cursor: Option<Hash>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub struct GetCompressedAccountsByOwnerResponse {
    pub context: Context,
    pub value: PaginatedAccountList,
}

pub async fn get_compressed_accounts_by_owner(
    conn: &DatabaseConnection,
    request: GetCompressedAccountsByOwnerRequest,
) -> Result<GetCompressedAccountsByOwnerResponse, PhotonApiError> {
    let context = Context::extract(conn).await?;
    let GetCompressedAccountsByOwnerRequest {
        owner,
        cursor,
        limit,
    } = request;

    let mut filter = accounts::Column::Owner
        .eq::<Vec<u8>>(owner.into())
        .and(accounts::Column::Spent.eq(false));

    if let Some(cursor) = cursor {
        filter = filter.and(accounts::Column::Hash.gt::<Vec<u8>>(cursor.into()));
    }
    let mut query_limit = PAGE_LIMIT;
    if let Some(limit) = limit {
        query_limit = limit.value();
    }

    let result = accounts::Entity::find()
        .order_by(accounts::Column::Hash, sea_orm::Order::Asc)
        .limit(query_limit)
        .filter(filter)
        .all(conn)
        .await?;

    let items = result
        .into_iter()
        .map(parse_account_model)
        .collect::<Result<Vec<Account>, PhotonApiError>>()?;

    let mut cursor = items.last().map(|u| u.hash.clone());
    if items.len() < query_limit as usize {
        cursor = None;
    }

    Ok(GetCompressedAccountsByOwnerResponse {
        context,
        value: PaginatedAccountList { items, cursor },
    })
}
