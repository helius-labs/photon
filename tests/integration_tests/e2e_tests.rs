use function_name::named;
use photon_indexer::api::method::get_compressed_accounts_by_owner::GetCompressedAccountsByOwnerRequest;
use photon_indexer::api::method::get_transaction_with_compression_info::get_transaction_helper;
use photon_indexer::common::typedefs::serializable_pubkey::SerializablePubkey;
use photon_indexer::ingester::index_block;
use solana_sdk::pubkey::Pubkey;

use crate::utils::*;
use insta::assert_json_snapshot;
use photon_indexer::api::method::utils::GetCompressedTokenAccountsByOwner;
use photon_indexer::common::typedefs::hash::Hash;
use photon_indexer::dao::generated::blocks;
use photon_indexer::ingester::typedefs::block_info::{BlockInfo, BlockMetadata};
use sea_orm::ColumnTrait;
use sea_orm::EntityTrait;
use sea_orm::QueryFilter;
use serial_test::serial;

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_e2e_mint_and_transfer(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    use std::str::FromStr;

    
    use photon_indexer::{
        api::method::{
            get_compression_signatures_for_token_owner::GetCompressionSignaturesForTokenOwnerRequest,
            get_multiple_compressed_account_proofs::HashList, utils::Limit,
        },
        common::typedefs::serializable_signature::SerializableSignature,
    };
    use solana_sdk::signature::Signature;

    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name.clone(),
        TestSetupOptions {
            network: Network::Localnet,
            db_backend,
        },
    )
    .await;

    let payer_pubkey =
        SerializablePubkey::try_from("Fg5TtbkvcqLedmwFqqbWty9NRqMKKMqy9zNLZZRQXSDc").unwrap();
    let bob_pubkey =
        SerializablePubkey::try_from("4rANawp9ahHmvaVi3CCeJSam8JTA4DDY3htJKnhYxbHN").unwrap();
    let charles_pubkey =
        SerializablePubkey::try_from("GyXNDaUUBwzYp24vEzang6SyCzcA8L3csZkiALqF6KRs").unwrap();

    let mint_tx =
        "2DcgLz1KwPaMoiRAAmYAirx4213swLbaSM1YYuCkbgzngq61XSWaFCRGuFoMgqW5htsEZRKBm6uQQPAPCXUVL8eH";
    let transfer_tx =
        "DFE7shhwCWd5P14NfA1ELSpd96FaJhBkjSet2Y7t7CtM4oLhP17AYKtyMBjZWi4hgFKvqET8zcAQUDuAC6sARfk";
    let transfer_txn2: &str =
        "5FDV6xrPRyfEUP2Kg7mwZVj34LVEfq9gsrquyFkvmmdPiiEn3Bhd8sTJTe6UhszNHLgwWfV6j9X4rsceL9mNNyRZ";
    let transfer_txn3 =
        "2pGQ7MgWuhrKpeSAzs58CXttKDiws33X68hQv4tRm6NUH6JCsCBk5uteNcjt2fXdToYhC6s8b6SDzoXV9CLJTo94";

    let txs = [mint_tx, transfer_tx, transfer_txn2, transfer_txn3];

    // HACK: We index a block so that API methods can fetch the current slot.
    index_block(
        &setup.db_conn,
        &BlockInfo {
            metadata: BlockMetadata {
                slot: 0,
                ..Default::default()
            },
            ..Default::default()
        },
    )
    .await
    .unwrap();

    for tx in txs {
        index_transaction(&setup, tx).await;
    }
    for (person, pubkey) in [
        ("bob", bob_pubkey.clone()),
        ("charles", charles_pubkey.clone()),
        ("payer", payer_pubkey.clone()),
    ] {
        let accounts = setup
            .api
            .get_compressed_token_accounts_by_owner(GetCompressedTokenAccountsByOwner {
                owner: pubkey.clone(),
                ..Default::default()
            })
            .await
            .unwrap();
        assert_json_snapshot!(format!("{}-{}-accounts", name.clone(), person), accounts);
        let proofs = setup
            .api
            .get_multiple_compressed_account_proofs(HashList(
                accounts
                    .value
                    .items
                    .iter()
                    .map(|x| x.account.hash.clone())
                    .collect(),
            ))
            .await
            .unwrap();
        assert_json_snapshot!(format!("{}-{}-proofs", name.clone(), person), proofs);

        let mut cursor = None;
        let limit = Limit::new(1).unwrap();
        let mut signatures = Vec::new();
        loop {
            let res = setup
                .api
                .get_compression_signatures_for_token_owner(
                    GetCompressionSignaturesForTokenOwnerRequest {
                        owner: pubkey.clone(),
                        cursor,
                        limit: Some(limit.clone()),
                    },
                )
                .await
                .unwrap()
                .value;
            signatures.extend(res.items);
            cursor = res.cursor;
            if cursor.is_none() {
                break;
            }
        }
        assert_json_snapshot!(
            format!("{}-{}-token-signatures", name.clone(), person),
            signatures
        );

        let token_balances = setup
                .api
                .get_compressed_token_balances_by_owner(photon_indexer::api::method::get_compressed_token_balances_by_owner::GetCompressedTokenBalancesByOwnerRequest {
                    owner: pubkey.clone(),
                    ..Default::default()
                })
                .await
                .unwrap();

        assert_json_snapshot!(
            format!("{}-{}-token-balances", name.clone(), person),
            token_balances
        );
    }
    for (txn_name, txn_signature) in [("mint", mint_tx), ("transfer", transfer_tx)] {
        let txn = cached_fetch_transaction(&setup, txn_signature).await;
        let txn_signature = SerializableSignature(Signature::from_str(txn_signature).unwrap());
        // Test get transaction
        let parsed_transaction: photon_indexer::api::method::get_transaction_with_compression_info::GetTransactionResponse = get_transaction_helper(&setup.db_conn, txn_signature, txn).await.unwrap();
        assert_json_snapshot!(
            format!("{}-{}-transaction", name.clone(), txn_name),
            parsed_transaction
        );
    }
}

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_lamport_transfers(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    use photon_indexer::api::method::get_multiple_compressed_account_proofs::HashList;

    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name.clone(),
        TestSetupOptions {
            network: Network::Localnet,
            db_backend,
        },
    )
    .await;

    let compress_tx =
        "3ZmBETHr8tHPr7iKdgW4NjTzcUCUHtSBFQD2JwSniShBV9fix6xb6FxRyR1cMRW83WerRYU5b1uxDLM3GqomCkYx";
    let transfer_tx1 =
        "64tDNpAnQHZsuiH8UG8vXDzXbwGHdghfLYerPdaVdgkQ2kSTE65kDT7ajzzhd2ZDW8nVoKYztnjxCaV8bWJxNCXR";
    let transfer_tx2 =
        "yK3j6TkABz2SVBtvsVwJV5ZoLv1tJBrBR8Z8JsosHQGbeizt7yx8Qx1Nva1ytcKa7hAXsT7R6WyQUiCH43Cwtk5";

    let payer_pubkey = SerializablePubkey(
        Pubkey::try_from("8wSvtG3dvL7Nso8E3wx6L9SwTRiZuR8ojRX5D6mwx7hJ").unwrap(),
    );
    let receiver_pubkey = SerializablePubkey(
        Pubkey::try_from("G1zCvDhtSdV6m92q3tXBEE4d531fBx1vD7EUpeeQZrNp").unwrap(),
    );
    let txs = [compress_tx, transfer_tx1, transfer_tx2];
    let number_of_indexes = [1, 2];

    for individually in [true, false] {
        reset_tables(&setup.db_conn).await.unwrap();
        for number in number_of_indexes.iter() {
            for _ in 0..*number {
                // HACK: We index a block so that API methods can fetch the current slot.
                index_block(
                    &setup.db_conn,
                    &BlockInfo {
                        metadata: BlockMetadata {
                            slot: 0,
                            ..Default::default()
                        },
                        ..Default::default()
                    },
                )
                .await
                .unwrap();

                match individually {
                    true => {
                        for tx in txs.iter() {
                            index_transaction(&setup, tx).await;
                        }
                    }
                    false => {
                        index_multiple_transactions(&setup, txs.to_vec()).await;
                    }
                }
            }
        }
        for (owner, owner_name) in [
            (payer_pubkey.clone(), "payer"),
            (receiver_pubkey.clone(), "receiver"),
        ] {
            let accounts = setup
                .api
                .get_compressed_accounts_by_owner(GetCompressedAccountsByOwnerRequest {
                    owner,
                    ..Default::default()
                })
                .await
                .unwrap();
            assert_json_snapshot!(
                format!("{}-{}-accounts", name.clone(), owner_name),
                accounts
            );
            let proofs = setup
                .api
                .get_multiple_compressed_account_proofs(HashList(
                    accounts
                        .value
                        .items
                        .iter()
                        .map(|x| x.hash.clone())
                        .collect(),
                ))
                .await
                .unwrap();
            assert_json_snapshot!(format!("{}-{}-proofs", name.clone(), owner_name), proofs);

            let signatures = setup
                .api
                .get_compression_signatures_for_owner(photon_indexer::api::method::get_compression_signatures_for_owner::GetCompressionSignaturesForOwnerRequest {
                    owner,
                    ..Default::default()
                })
                .await
                .unwrap();

            let limit = photon_indexer::api::method::utils::Limit::new(1).unwrap();
            let mut cursor = None;
            let mut paginated_signatures = Vec::new();
            loop {
                let res = setup
                    .api
                    .get_compression_signatures_for_owner(photon_indexer::api::method::get_compression_signatures_for_owner::GetCompressionSignaturesForOwnerRequest {
                        owner: owner.clone(),
                        cursor,
                        limit: Some(limit.clone()),
                    })
                    .await
                    .unwrap()
                    .value;
                paginated_signatures.extend(res.items);
                cursor = res.cursor;
                if cursor.is_none() {
                    break;
                }
            }
            assert_eq!(signatures.value.items, paginated_signatures);
            assert_json_snapshot!(
                format!("{}-{}-signatures", owner_name, name.clone()),
                signatures
            );

            let balance = setup
                .api
                .get_compressed_balance_by_owner(photon_indexer::api::method::get_compressed_balance_by_owner::GetCompressedBalanceByOwnerRequest {
                    owner,
                    ..Default::default()
                })
                .await
                .unwrap();

            assert_eq!(
                accounts
                    .value
                    .items
                    .iter()
                    .map(|x| x.lamports.0)
                    .sum::<u64>(),
                balance.value.0,
            );
        }
    }
}

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_index_block_metadata(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name.clone(),
        TestSetupOptions {
            network: Network::Mainnet,
            db_backend,
        },
    )
    .await;

    let slot = 254170887;
    let block = cached_fetch_block(&setup, slot).await;
    index_block(&setup.db_conn, &block).await.unwrap();
    let filter = blocks::Column::Slot.eq(block.metadata.slot);

    let block_model = blocks::Entity::find()
        .filter(filter)
        .one(setup.db_conn.as_ref())
        .await
        .unwrap()
        .unwrap();

    assert_eq!(block_model.slot, 254170887);
    assert_eq!(block_model.parent_slot, 254170886);
    assert_eq!(
        Hash::try_from(block_model.parent_blockhash)
            .unwrap()
            .to_string(),
        "9BMHTdybcGah8PWtCzu8tVFDBXmiEHZZDf3FaZ651Nf"
    );
    assert_eq!(
        Hash::try_from(block_model.blockhash).unwrap().to_string(),
        "5GG5pzTbH6KgZM54M9XWfBNmBupQuZXdQxoZRRQXHcpM"
    );
    assert_eq!(block_model.block_height, 234724352);
    assert_eq!(block_model.block_time, 1710441678);

    // Verify that we don't get an error if we try to index the same block again
    index_block(&setup.db_conn, &block).await.unwrap();
    assert_eq!(setup.api.get_indexer_slot().await.unwrap().0, slot);

    // Verify that get_indexer_slot() gets updated a new block is indexed.
    let block = cached_fetch_block(&setup, slot + 1).await;
    index_block(&setup.db_conn, &block).await.unwrap();
    assert_eq!(setup.api.get_indexer_slot().await.unwrap().0, slot + 1);
}
