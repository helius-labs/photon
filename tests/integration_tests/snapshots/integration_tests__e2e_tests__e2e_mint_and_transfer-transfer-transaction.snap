---
source: tests/integration_tests/e2e_tests.rs
expression: parsed_transaction
---
{
  "transaction": {
    "slot": 53,
    "transaction": [
      "AjfneV2HbG23LQexYuLHAoF9wMZsj0IOpeXvDxNGbqdFXUFjTJk/kSNb1bGUTR68GLddy/GdsASqEcWcYBSiqQ27gMDUdcZjuKumQ60bMqQsJaWyihPVqHeZKaCFzXeiLUlIyILzjknpoQjvI5yuVLJ9X55lJIU4BsX9HVJl1FoEgAIBCQ129wO8WIKG6eY3uWLE8oiLV+uv4g4bZ8FJk67jITSz9GFZsjIhiI/k4Jf83FMyo4EszCS+vrMsRiha6mJgMIKeRE19IEyAPbQBz0UseXaZEbO3cyKjf2bWAVevsV8xQUUtaxlkRkeJL9L43N8LzM1Z+DDTqAsZ255jL3/4jmvEbgMGRm/lIRcy/+ytunLDm+e8jOW7xfcSayxDmzpAAAAA9YrSxwEeKNSgRrWi/d8zJ5QeTRw+kg0Uqpn2e2KyQR+0pUDsg7fVlu64SNoTe62H9ztjOse7tyCQDbQ9bkxgnO76I9ijWlJIp9HE49KtWf+OFMeWkNc9pCIvteTimdARkc7UXi57fp4+O8qFD/X5tZdG4r6MA+E6LKGyLVgSOLkLvA/Au0fKL3TEES6UqxPPo8Y05dwX6ssDzRojzX54fJQlr738AEklZ5z/PVi137S1ja67KHPEXpbltmRVKWz4rFbVF+VphmG2mUNlOfY6oA8e5Sx7wpJM4c+seQXacoYAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAALruaUcu7+H8CMsmyTtJGWXL9h4FbXnCPweCUcdCDZBMAgQABQJAQg8ABQ8AAQYHCAkKCwUFBQUMAgOgAqM0yOeMA0W6FAEAAAGLZI542l/AiGL98WqbD7cn8gKT/fTrAAaqr+iWPzYnxCORzyU0mBJGa//k3L+m8rNQcs8+gRPW9NG1vvUFmcFNDCeNlVXlK0n/c3Z7OEpg7GWTdn2aS8DxxnqWzCeVmz8P8yU9A/2KsEvBmfHFDBul/YMCb8k4rqiyBDllDYuxrtzpL6rbPGdKqE1hHsDJDmlITXcq5npf614uZg2kjr4/AAEAAADoAwAAAAAAAAAAAAABAAAAAAEAAgAAAGFZsjIhiI/k4Jf83FMyo4EszCS+vrMsRiha6mJgMIKeLAEAAAAAAAAAAIiI6cSyl5N451k0XO5yRDnU7dhk5Bt4Zgn4ddmv4+AIvAIAAAAAAAAAAAAAAAA=",
      "base64"
    ],
    "meta": {
      "err": null,
      "status": {
        "Ok": null
      },
      "fee": 10000,
      "preBalances": [
        996478932,
        1000000000,
        9496836668,
        2293180800,
        1,
        1141440,
        0,
        1141440,
        1392000,
        1141440,
        0,
        1141440,
        1
      ],
      "postBalances": [
        996463556,
        1000000000,
        9496837044,
        2293185800,
        1,
        1141440,
        0,
        1141440,
        1392000,
        1141440,
        0,
        1141440,
        1
      ],
      "innerInstructions": [
        {
          "index": 1,
          "instructions": [
            {
              "programIdIndex": 7,
              "accounts": [
                0,
                6,
                8,
                9,
                10,
                11,
                5,
                7,
                7,
                12,
                7,
                2,
                3
              ],
              "data": "AdP3Yq8kJWgSqv2pJsUz3UKGRejVMHRMBy5wAxV1gWKVRCGLw36zd5MGXGLt9HF9TXy3J6wcYq8x2x3CbF3C3kfvmzVeeA4ZK2mJufjHeFUwo6G7Sexhxm18FXAC5ZnuxPNHaFWfdL4ustZADebC1ECWJCjndiZdcZTLzeSkcyv4xApWiEYxqaRXPBBgawgzAft8WAVZjR7cSHioebmxx3MMn6YdAdfXHqDCRr8jKLnJQ5dKNFbMoGsMRneq6tPGykjtcK1ZBcB5d29sbgFNdFvxzK95Rh2brQcTcPXtwA389bgHRL5MAtRvuoNsqyVWwVzDjF14tu8PRPrt99pW33SaCAGBjHTtnJ1qheX3NP3w3gs4gBtdVxbSsQs5X6tcdV8T3UZiqtvrELixFwYj8Zo4oUHXrn63ovsftFfdmpfSzSebANLw34Jb7tdCD2nFnT5PVrPw9d8dBduutho6mJAeyJbL2bjHU1xj2LThpMDKC2PEK8XSCVqrhYpouMe2EFkuF3w6KxcKw1kJ1ZZdoQdqaknN6EsmwFqHgDgggLRePwPFh3jNYkfyw7xwQEccPsHRmfk5w8qY2DbAgksUhLzmG8L6kiN1d3zSGvogvRf6BpeiayvPRKvDApREosAgqAXhg8nfPhgNifwc7p3oLZwtwLTACcZk1TVYXCv8KfTnMBT1Sa9V1NGBZoxYsQzhFg86XWcPjjk6hxd5zU8D4p5koMdgDgC2ChMntqSxYyt6NyeueTyub7dtHWLESXZopQsXfD1ALcgfsfqFYQcnnTVnnxWv8QN7xDzkRGRkeJo2GT2p2kAMqkVtcFBRjT796ePdHseA4rwhkpb7fjFmGfUM7YG7UsP41uh949366nbukz79QefSdceEhE93xnxi62yv7xYbAYEF3n5fqAJXyB4HPXVjQz1UKUFYAoKtbgVFFX3W43MUijcAPZaf",
              "stackHeight": 2
            },
            {
              "programIdIndex": 11,
              "accounts": [
                0,
                10,
                8,
                12,
                3,
                2
              ],
              "data": "WiZw2hSp2LSxinAV4MZFrZhyqHeSCxYy6pTQBzotg13s2HgckUNVRmGvRp89",
              "stackHeight": 3
            },
            {
              "programIdIndex": 12,
              "accounts": [
                0,
                3
              ],
              "data": "3Bxs4PckVVt51W8w",
              "stackHeight": 3
            },
            {
              "programIdIndex": 11,
              "accounts": [
                0,
                10,
                8,
                12,
                2
              ],
              "data": "7UWNMdwFJEWBQPSoUMKtr3bukccHKfYDfeQQkEJTkRR52MqsyybbmeLhdW4rUSQzfgw6LDbMb3emjEZMC6T8y8yf2kPoXgNW4jarNgTDhrZ",
              "stackHeight": 3
            },
            {
              "programIdIndex": 12,
              "accounts": [
                0,
                2
              ],
              "data": "3Bxs4LvrAdJLFH6X",
              "stackHeight": 4
            },
            {
              "programIdIndex": 9,
              "accounts": [],
              "data": "3oLtGWPG6EsRhRwBpFEBZbQCbrDkBnwUPsiSYzoBL5hCPaTBxFQKqexZftuBgWrEtmTPnxZcv3dB5FcuDcv2rNEWXeBwXAx6QDV2Zbo7DPk2XVmQ3ZsbAoPsk9pxZSHhNK14phsibkkT1Vr9XU1U7pLr86wTDSMpRXscMnWqmb5YpaG9XF1VMHrWW8QJdd5BTyc9FQCYFYQC8ctqhcfkWWn6iFh6bQsYScz7cARKVqRD6PstzkbVWiuq9zm4V3Bht6hsC4aBWfXGGjwCvvs4bN41fAbCQaoZtGGpRnU7dNRnSh3vayYyZBnwPyzpuL6QYn1bjHDb2rDqToVbog9Jsq3MV5D7Xod9AB4ua4Wke7mBZKy2nEwwEKgm36ffxbx4XLUZuQ4MTSQHiSPyy2JC3RsSHqR1yCUneDBWzmHdUCx66KNgQGcBpnc7qoPqhcfRGSCgPkcoLsx6AzovJduWg85r4ijbWmrscvNdqj1W2SbyBRPWsiB7jPaGogV1xdHBGoU4bj8Tkd1mtsLDX7oa7uz1KSXABgRrdtyZaRsL66cCiv8L4bd2hjo6St1e86SPGsdADi4SFHUmfRTqVZHXcPWPDC7QTUuNsTHSTjKKGuG4VUz4sanxm8Woq8wa2EaUtoDckp6ERaNznMKTxTh9LZ4ga4Qji67BvXQsbe62Z6bF1CkkbSCiEHT7AibYnGnzuiyiyiQ3LsLYQ3QKKouhioayWNEtwFZaX1BykDjGR3uytiWWBYUzhyXqqJhpdPH3B84PFfHPUQgiWgyNm5",
              "stackHeight": 3
            }
          ]
        }
      ],
      "logMessages": [
        "Program ComputeBudget111111111111111111111111111111 invoke [1]",
        "Program ComputeBudget111111111111111111111111111111 success",
        "Program HXVfQ44ATEi9WBKLSCCwM54KokdkzqXci9xCQ7ST9SYN invoke [1]",
        "Program log: Instruction: Transfer",
        "Program H5sFv8VwWmjxHYS2GB4fTDsK7uTtnRT4WiixtHrET3bN invoke [2]",
        "Program log: Instruction: InvokeCpi",
        "Program CbjvJc1SNx1aav8tU49dJGHu8EUdzQJSMtkjDmV8miqK invoke [3]",
        "Program log: Instruction: InsertIntoNullifierQueues",
        "Program CbjvJc1SNx1aav8tU49dJGHu8EUdzQJSMtkjDmV8miqK consumed 9341 of 816825 compute units",
        "Program CbjvJc1SNx1aav8tU49dJGHu8EUdzQJSMtkjDmV8miqK success",
        "Program 11111111111111111111111111111111 invoke [3]",
        "Program 11111111111111111111111111111111 success",
        "Program CbjvJc1SNx1aav8tU49dJGHu8EUdzQJSMtkjDmV8miqK invoke [3]",
        "Program log: Instruction: AppendLeavesToMerkleTrees",
        "Program 11111111111111111111111111111111 invoke [4]",
        "Program 11111111111111111111111111111111 success",
        "Program CbjvJc1SNx1aav8tU49dJGHu8EUdzQJSMtkjDmV8miqK consumed 51939 of 790356 compute units",
        "Program CbjvJc1SNx1aav8tU49dJGHu8EUdzQJSMtkjDmV8miqK success",
        "Program noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV invoke [3]",
        "Program noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV consumed 38 of 734927 compute units",
        "Program noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV success",
        "Program H5sFv8VwWmjxHYS2GB4fTDsK7uTtnRT4WiixtHrET3bN consumed 226867 of 961128 compute units",
        "Program H5sFv8VwWmjxHYS2GB4fTDsK7uTtnRT4WiixtHrET3bN success",
        "Program HXVfQ44ATEi9WBKLSCCwM54KokdkzqXci9xCQ7ST9SYN consumed 266570 of 999850 compute units",
        "Program HXVfQ44ATEi9WBKLSCCwM54KokdkzqXci9xCQ7ST9SYN success"
      ],
      "preTokenBalances": [],
      "postTokenBalances": [],
      "rewards": [],
      "loadedAddresses": {
        "writable": [],
        "readonly": []
      },
      "computeUnitsConsumed": 266720
    },
    "version": 0,
    "blockTime": 1717594775
  },
  "compressionInfo": {
    "closedAccounts": [
      {
        "account": {
          "hash": "sKXVGdDgbogNTGsoa1eWGcE4u7M7PbdoTYBD6L739yB",
          "address": null,
          "data": {
            "discriminator": 2,
            "data": "3Okvqts8Z0qoTWEewMkOaUhNdyrmel/rXi5mDaSOvj9hWbIyIYiP5OCX/NxTMqOBLMwkvr6zLEYoWupiYDCCnugDAAAAAAAAAAAAAAAAAAAAAAA=",
            "dataHash": "Ta3F7qfK5vNRy1zPB1ifmxeHM1ZfBaqxqUmBRWVCX8Y"
          },
          "owner": "HXVfQ44ATEi9WBKLSCCwM54KokdkzqXci9xCQ7ST9SYN",
          "lamports": 0,
          "tree": "5bdFnXU47QjzGpzHfXnxcEi5WXyxzEAZzd1vrE39bf1W",
          "leafIndex": 0,
          "seq": 1,
          "slotCreated": 0
        },
        "optionalTokenData": {
          "mint": "FsLvdEqi6xTaxSDZ6mzCRhWGRN6W7c7n8E4SmPix6YM4",
          "owner": "7Z1siJsXDpSgwyXsErgdrhbLNHGuqoivfjbycnUyxWX7",
          "amount": 1000,
          "delegate": null,
          "state": "initialized",
          "isNative": null,
          "delegatedAmount": 0
        }
      }
    ],
    "openedAccounts": [
      {
        "account": {
          "hash": "3E79jMkJGfK5dvJ5uFCVZUUYQt62N23UUQjtLqL6Xup8",
          "address": null,
          "data": {
            "discriminator": 2,
            "data": "3Okvqts8Z0qoTWEewMkOaUhNdyrmel/rXi5mDaSOvj9hWbIyIYiP5OCX/NxTMqOBLMwkvr6zLEYoWupiYDCCniwBAAAAAAAAAAAAAAAAAAAAAAA=",
            "dataHash": "ipgQY2rAKLuR54ZgkkLtivtr1qQhd5CGado3R3WQxhB"
          },
          "owner": "HXVfQ44ATEi9WBKLSCCwM54KokdkzqXci9xCQ7ST9SYN",
          "lamports": 0,
          "tree": "5bdFnXU47QjzGpzHfXnxcEi5WXyxzEAZzd1vrE39bf1W",
          "leafIndex": 1,
          "seq": 2,
          "slotCreated": 53
        },
        "optionalTokenData": {
          "mint": "FsLvdEqi6xTaxSDZ6mzCRhWGRN6W7c7n8E4SmPix6YM4",
          "owner": "7Z1siJsXDpSgwyXsErgdrhbLNHGuqoivfjbycnUyxWX7",
          "amount": 300,
          "delegate": null,
          "state": "initialized",
          "isNative": null,
          "delegatedAmount": 0
        }
      },
      {
        "account": {
          "hash": "2wN5qe6d9ptX8VkSN7QYS8qAnRMMmVhmFQET29BNe5PP",
          "address": null,
          "data": {
            "discriminator": 2,
            "data": "3Okvqts8Z0qoTWEewMkOaUhNdyrmel/rXi5mDaSOvj+IiOnEspeTeOdZNFzuckQ51O3YZOQbeGYJ+HXZr+PgCLwCAAAAAAAAAAAAAAAAAAAAAAA=",
            "dataHash": "3kc71bVFqhxuPprfvg2ZeVi7oTzLdoSZrCgRKhBXJ7RH"
          },
          "owner": "HXVfQ44ATEi9WBKLSCCwM54KokdkzqXci9xCQ7ST9SYN",
          "lamports": 0,
          "tree": "5bdFnXU47QjzGpzHfXnxcEi5WXyxzEAZzd1vrE39bf1W",
          "leafIndex": 2,
          "seq": 3,
          "slotCreated": 53
        },
        "optionalTokenData": {
          "mint": "FsLvdEqi6xTaxSDZ6mzCRhWGRN6W7c7n8E4SmPix6YM4",
          "owner": "AByXwz518ntDjhTaMr6Z65pgde3gy7t4tZ6htdPGkjhh",
          "amount": 700,
          "delegate": null,
          "state": "initialized",
          "isNative": null,
          "delegatedAmount": 0
        }
      }
    ]
  }
}
